API/Manager/ServerManager.cs

```diff
@@ -111,5 +127,65 @@ public IEnumerable<AudioPreset> GetAudioPresets()
         {
             return _transcoder.GetAvailableAudioPresets();
         }
+
+        public void KeepAlive(KeepAliveRequest request)
+        {
+            lock (TranscoderCache)
+            {
+                var cacheObject = TranscoderCache.FirstOrDefault(item => item.TranscodedVideoUri == request.TranscodedVideoUri &&
+                    item.AudioPresetID == request.AudioPreset &&
+                    item.VideoPresetID == request.VideoPreset);
+                if(cacheObject != null)
+                {
+                cacheObject.KeepAliveTimeStamp = DateTime.Now;
+                }
+                //TODO: Response to Client? Keepalive Invalid, Video stopped to transcode
+            }
+
+        }
+
+        private Task CheckTranscodingCache()
+        {
+
+            Task checkCache = new Task(() =>
+            {
+
+                while (true)
+                {
+                    
+                    //Locked for operations
+                    lock (TranscoderCache)
+                    {
+                        if (TranscoderCache.Count == 0)
+                        continue;
+
+                        //Cant delete IEnumerable Entries while in for each
+                        var IterationList = new List<TranscoderCachingObject>();
+                        IterationList.InsertRange(0, TranscoderCache);
+
+                        foreach (var video in IterationList)
+                        {
+                            if ((DateTime.Now - video.KeepAliveTimeStamp) > TimeSpan.FromMinutes(1))
+                            {
+                                _logger.LogInformation($"Keepalive for Video {video.VideoSourceUri} with videoPreset {video.VideoPresetID} and audioPreset {video.AudioPresetID} expired");
+                                //Deactivate Token to stop Process
+                                video.CancellationTokenSource.Cancel();
+                            }
+                            if (video.CancellationTokenSource.IsCancellationRequested)
+                            {
+                                //Delete Entry in real Cache
+                                _logger.LogInformation($"Process for Video {video.VideoSourceUri} with videoPreset {video.VideoPresetID} and audioPreset {video.AudioPresetID} removed from Cache");
+                                TranscoderCache.Remove(video);
+                                video.CancellationTokenSource.Dispose();
+                            }
+                        }
+                        IterationList = null;
+                    }
+                    Thread.Sleep(5000);
```

Comment:
Man könnte hier einen Scheduler verwenden statt thread.sleep, wird aber wohl keinen Unterschied machen. Zum Chache hätte man auch Microsoft.Extensions.Caching.Memory verwenden können, da könnte mach max. size und eine expire time angeben.
So passt ja auch alles deswegen von mir ok.

Reply:
Ich glaube dass es ziemlich aufwändig wird, wenn man statt dem Default Scheduler einen eigenen verwendet. (Die ThreadScheduler Klasse ist abstrakt und muss daher erst eigens implementiert werden). So hab ich das zumindest verstanden. Da wir kaum Hintergrundprozesse haben dürfte Thread.Sleep im Moment nicht problematisch werden, da der Threadpool dadurch nur gering beeinflusst wird.

Zu Microsoft.Extensions.Caching.Memory kann ich recht wenig sagen. Diese Klasse kann aber immernoch im Nachhinein in Betracht gezogen werden.